src/js/plugin.js

```diff
@@ -0,0 +1,445 @@
+ */
+const BASE_PLUGIN_NAME = 'plugin';
+
+/**
+ * The key on which a player's active plugins cache is stored.
+ *
+ * @private
+ * @type {string}
+ */
+const PLUGIN_CACHE_KEY = 'activePlugins_';
+
+/**
+ * Stores registered plugins in a private space.
+ *
+ * @private
+ * @type {Object}
+ */
+const pluginStorage = {};
+
+/**
+ * Reports whether or not a plugin has been registered.
+ *
+ * @private
+ * @param  {string} name
+ *         The name of a plugin.
+ *
+ * @return {boolean}
+ *         Whether or not the plugin has been registered.
+ */
+const pluginExists = (name) => pluginStorage.hasOwnProperty(name);
+
+/**
+ * Get a single registered plugin by name.
+ *
+ * @private
+ * @param  {string} name
+ *         The name of a plugin.
+ *
+ * @return {Function|undefined}
+ *         The plugin (or undefined).
+ */
+const getPlugin = (name) => pluginExists(name) ? pluginStorage[name] : undefined;
+
+/**
+ * Marks a plugin as "active" on a player.
+ *
+ * Also, ensures that the player has an object for tracking active plugins.
+ *
+ * @private
+ * @param  {Player} player
+ *         A Video.js player instance.
+ *
+ * @param  {string} name
+ *         The name of a plugin.
+ */
+const markPluginAsActive = (player, name) => {
+  player[PLUGIN_CACHE_KEY] = player[PLUGIN_CACHE_KEY] || {};
+  player[PLUGIN_CACHE_KEY][name] = true;
+};
+
+/**
+ * Takes a basic plugin function and returns a wrapper function which marks
+ * on the player that the plugin has been activated.
+ *
+ * @private
+ * @param  {string} name
+ *         The name of the plugin.
+ *
+ * @param  {Function} plugin
+ *         The basic plugin.
+ *
+ * @return {Function}
+ *         A wrapper function for the given plugin.
+ */
+const createBasicPlugin = (name, plugin) => function() {
+  const instance = plugin.apply(this, arguments);
+
+  markPluginAsActive(this, name);
+
+  // We trigger the "pluginsetup" event on the player regardless, but we want
+  // the hash to be consistent with the hash provided for advanced plugins.
+  // The only potentially counter-intuitive thing here is the `instance` is the
+  // value returned by the `plugin` function.
+  this.trigger('pluginsetup', {name, plugin, instance});
+  return instance;
+};
+
+/**
+ * Takes a plugin sub-class and returns a factory function for generating
+ * instances of it.
+ *
+ * This factory function will replace itself with an instance of the requested
+ * sub-class of Plugin.
+ *
+ * @private
+ * @param  {string} name
+ *         The name of the plugin.
+ *
+ * @param  {Plugin} PluginSubClass
+ *         The advanced plugin.
+ *
+ * @return {Function}
+ *         A factory function for the plugin sub-class.
+ */
+const createPluginFactory = (name, PluginSubClass) => {
+
+  // Add a `name` property to the plugin prototype so that each plugin can
+  // refer to itself by name.
+  PluginSubClass.prototype.name = name;
+
+  return function(...args) {
+    const instance = new PluginSubClass(...[this, ...args]);
```

Comment:
do we want to create a new plugin each time `player.pluginName()` is called? Don't we want to return the existing instance if one already exists that has not been disposed?